File: Backend/StrategyEngine/Simulation/Stochastic.fs
module StrategyEngine.Simulation.Stochastic

open System

// ... (NormalRandom stays the same) ...
type NormalRandom(seed: int) =
    let rnd = System.Random(seed)
    let mutable hasSpare = false
    let mutable spare = 0.0
    member _.Next() =
        if hasSpare then
            hasSpare <- false
            spare
        else
            let mutable u = 0.0
            let mutable v = 0.0
            let mutable s = 0.0
            while s >= 1.0 || s = 0.0 do
                u <- rnd.NextDouble() * 2.0 - 1.0
                v <- rnd.NextDouble() * 2.0 - 1.0
                s <- u * u + v * v
            let mul = Math.Sqrt(-2.0 * Math.Log(s) / s)
            hasSpare <- true
            spare <- v * mul
            u * mul

module LinearAlgebra =
    
    // CHANGED: Input float[][] instead of float[,]
    let choleskyDecomposition (matrix: float[][]) : float[][] =
        let n = matrix.Length
        // Initialize jagged array
        let result = Array.init n (fun _ -> Array.zeroCreate n)
        
        for i in 0 .. n - 1 do
            for j in 0 .. i do
                let mutable sum = 0.0
                for k in 0 .. j - 1 do
                    sum <- sum + result.[i].[k] * result.[j].[k]
                
                if i = j then
                    let diff = matrix.[i].[i] - sum
                    if diff <= 0.0 then 
                        failwith "Matrix is not positive definite"
                    result.[i].[j] <- Math.Sqrt(diff)
                else
                    result.[i].[j] <- (1.0 / result.[j].[j]) * (matrix.[i].[j] - sum)
        result

    // CHANGED: Input float[][]
    let multiplyMatrixVector (matrix: float[][]) (vector: float[]) : float[] =
        let n = matrix.Length
        let result = Array.zeroCreate n
        
        for i in 0 .. n - 1 do
            let mutable sum = 0.0
            for j in 0 .. i do 
                sum <- sum + matrix.[i].[j] * vector.[j]
            result.[i] <- sum
        result

File: Frontend/src/assets/fable_build/Simulation/Stochastic.js
import { seeded } from "../fable_modules/fable-library-js.4.27.0/Random.js";
import { class_type } from "../fable_modules/fable-library-js.4.27.0/Reflection.js";
import { setItem, item, initialize } from "../fable_modules/fable-library-js.4.27.0/Array.js";

export class NormalRandom {
    constructor(seed) {
        this.rnd = seeded(seed);
        this.hasSpare = false;
        this.spare = 0;
    }
}

export function NormalRandom_$reflection() {
    return class_type("StrategyEngine.Simulation.Stochastic.NormalRandom", undefined, NormalRandom);
}

export function NormalRandom_$ctor_Z524259A4(seed) {
    return new NormalRandom(seed);
}

export function NormalRandom__Next(_) {
    if (_.hasSpare) {
        _.hasSpare = false;
        return _.spare;
    }
    else {
        let u = 0;
        let v = 0;
        let s = 0;
        while ((s >= 1) ? true : (s === 0)) {
            u = ((_.rnd.NextDouble() * 2) - 1);
            v = ((_.rnd.NextDouble() * 2) - 1);
            s = ((u * u) + (v * v));
        }
        const mul = Math.sqrt((-2 * Math.log(s)) / s);
        _.hasSpare = true;
        _.spare = (v * mul);
        return u * mul;
    }
}

export function LinearAlgebra_choleskyDecomposition(matrix) {
    const n = matrix.length | 0;
    const result = initialize(n, (_arg) => (new Float64Array(n)));
    for (let i = 0; i <= (n - 1); i++) {
        for (let j = 0; j <= i; j++) {
            let sum = 0;
            for (let k = 0; k <= (j - 1); k++) {
                sum = (sum + (item(k, item(i, result)) * item(k, item(j, result))));
            }
            if (i === j) {
                const diff = item(i, item(i, matrix)) - sum;
                if (diff <= 0) {
                    throw new Error("Matrix is not positive definite");
                }
                item(i, result)[j] = Math.sqrt(diff);
            }
            else {
                item(i, result)[j] = ((1 / item(j, item(j, result))) * (item(j, item(i, matrix)) - sum));
            }
        }
    }
    return result;
}

export function LinearAlgebra_multiplyMatrixVector(matrix, vector) {
    const n = matrix.length | 0;
    const result = new Float64Array(n);
    for (let i = 0; i <= (n - 1); i++) {
        let sum = 0;
        for (let j = 0; j <= i; j++) {
            sum = (sum + (item(j, item(i, matrix)) * item(j, vector)));
        }
        setItem(result, i, sum);
    }
    return result;
}

